lib/elixir/lib/kernel.ex

```diff
@@ -5105,21 +5127,63 @@ defmodule Kernel do
   defmacro sigil_U(datetime_string, modifiers)
 
   defmacro sigil_U({:<<>>, _, [string]}, []) do
-    Macro.escape(datetime_from_utc_iso8601!(string))
+    {{:ok, {year, month, day, hour, minute, second, microsecond}, offset}, calendar} =
+      parse_with_calendar!(string, :parse_utc_datetime, "UTC DateTime")
+
+    if offset != 0 do
+      raise ArgumentError,
+            "cannot parse #{inspect(string)} as UTC DateTime for #{inspect(calendar)}, reason: :non_utc_offset"
+    end
+
+    to_calendar_struct(DateTime,
+      calendar: calendar,
+      year: year,
+      month: month,
+      day: day,
+      hour: hour,
+      minute: minute,
+      second: second,
+      microsecond: microsecond,
+      time_zone: "Etc/UTC",
+      zone_abbr: "UTC",
+      utc_offset: 0,
+      std_offset: 0
+    )
   end
 
-  defp datetime_from_utc_iso8601!(string) do
-    case DateTime.from_iso8601(string) do
-      {:ok, utc_datetime, 0} ->
-        utc_datetime
+  defp parse_with_calendar!(string, fun, context) do
+    {calendar, string} = extract_calendar(string)
+    {calendar |> apply(fun, [string]) |> maybe_raise!(calendar, context, string), calendar}
+  end
 
-      {:ok, _datetime, _offset} ->
-        raise ArgumentError,
-              "cannot parse #{inspect(string)} as UTC datetime, reason: :non_utc_offset"
+  defp extract_calendar(string) do
+    case :binary.split(string, " ", [:global]) do
+      [_] -> {Calendar.ISO, string}
+      parts -> maybe_atomize_calendar(List.last(parts), string)
+    end
+  end
 
-      {:error, reason} ->
-        raise ArgumentError,
-              "cannot parse #{inspect(string)} as UTC datetime, reason: #{inspect(reason)}"
+  defp maybe_atomize_calendar(<<alias, _::binary>> = last_part, string)
+       when alias >= ?A and alias <= ?Z do
+    string = binary_part(string, 0, byte_size(string) - byte_size(last_part) - 1)
+    {String.to_atom("Elixir." <> last_part), string}
+  end
```

Comment:
If I'm following this correctly, an alias wouldn't be expanded here. Meaning that:
```
defmodule MyModule do
  alias Calendar.Holocene
  ~D[12345-01-01 Holocene]
end
```
Would raise because `Holocene` wouldn't be expanded to `Calendar.Holocene`.  I think it would be good if an attempt to expand an alias was possible.

Reply:
I don't want to support aliases because it means if we copy and paste the code, then the inspect representation won't work unless the alias is the same and that literally never happens with any other inspect representation. :)

Reply:
Hmmm, my understanding was that the inspect format is always the full module name and therefore would happily parse on roundtrip because it wouldn't expand. But since its a string being parsed it could also be an unexpected side effect if expanded.

Reply:
Right, if you copy from the inspected result it would be the full name, but I am saying if you copy it from the source code. Basically, I expected `MyAlias` to be generally dependent on its module, but I wouldn't expect it to happen from inside a sigil (which are often seen as strings).

Reply:
Yes, I agree thats a better priority: Sigil's should be copyable without concern for context or side affects.